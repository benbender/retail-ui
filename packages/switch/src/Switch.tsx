import clsx from 'clsx'
import React from 'react'

type Ref = HTMLInputElement
type ReactInputProps = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  'checked' | 'disabled'
>

export type SwitchProps = {
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
  isChecked?: boolean
  isDisabled?: boolean
}

export const Switch = React.forwardRef<Ref, ReactInputProps & SwitchProps>(
  (props, ref) => {
    const { className, onChange, isChecked, isDisabled, ...rest } = props

    // class="
    const cls = clsx(
      'relative inline-block align-middle cursor-pointer select-none bg-transparent focus-within:shadow-outline',
    )
    // "

    // class="
    const trackCls = clsx(
      `w-12 h-6 bg-purple-600 dark:bg-gray-600 rounded-full shadow-inner`,
      isChecked && `transform transition-colors bg-green-500`,
      isDisabled && `bg-purple-500`,
    )
    // "

    // class="
    const thumbCls = clsx(
      'transition-all duration-300 ease-in-out absolute top-0 left-0 w-6 h-6 bg-white border-2 border-purple-600 rounded-full shadow',
      isChecked && `transform translate-x-full border-green-500`,
      isDisabled && `bg-purple-100 border-purple-500`,
    )
    // "

    return (
      // The label is required for the trick
      <label className={cls}>
        <input
          ref={ref}
          type="checkbox"
          className="hidden"
          onChange={onChange}
          {...rest}
        />
        <div className={trackCls}></div>
        <div className={thumbCls}></div>
      </label>
    )
  },
)

Switch.displayName = 'Switch'
