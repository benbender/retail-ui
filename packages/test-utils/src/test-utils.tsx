import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
