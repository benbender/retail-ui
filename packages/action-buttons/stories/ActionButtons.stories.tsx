import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
