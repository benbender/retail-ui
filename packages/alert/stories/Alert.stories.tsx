import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
