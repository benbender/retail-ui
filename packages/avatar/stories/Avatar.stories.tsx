import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
