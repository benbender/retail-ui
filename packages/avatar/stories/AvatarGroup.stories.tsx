import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
