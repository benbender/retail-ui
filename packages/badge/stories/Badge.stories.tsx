import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
