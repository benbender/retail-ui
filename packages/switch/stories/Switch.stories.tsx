import * as React from 'react'

import { Switch, SwitchProps } from '../src'

const meta = {
  title: 'Retail-UI/Switch',
  component: Switch,
}

export default meta

const Template = (args: SwitchProps) => (
  <Switch {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
