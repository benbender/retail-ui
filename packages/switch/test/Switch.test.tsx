import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Switch from '../stories/Switch.stories'

describe('Switch', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Switch.Simple />)
    expect(asFragment()).toMatchSnapshot()
  })
})
