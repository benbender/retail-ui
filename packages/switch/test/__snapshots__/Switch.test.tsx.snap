// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders correctly 1`] = `
<DocumentFragment>
  <label
    class="relative inline-block align-middle cursor-pointer select-none bg-transparent focus-within:shadow-outline"
  >
    <input
      class="hidden"
      data-testid="Retail-UI/Switch"
      type="checkbox"
    />
    <div
      class="w-12 h-6 bg-purple-600 dark:bg-gray-600 rounded-full shadow-inner"
    />
    <div
      class="transition-all duration-300 ease-in-out absolute top-0 left-0 w-6 h-6 bg-white border-2 border-purple-600 rounded-full shadow"
    />
  </label>
</DocumentFragment>
`;
